src/text/text_formatter.rs

```diff
@@ -550,9 +549,11 @@ mod formatter_test {
     fn test_format_struct() -> IonResult<()> {
         formatter(
             |ivf| {
-                ivf.format_struct(&Struct::from_iter(
-                    vec![("greetings", Element::from(Value::String("hello".into())))].into_iter(),
-                ))
+                ivf.format_struct(
+                    &Struct::builder()
+                        .with_field("greetings", "hello")
+                        .build_struct(),
```

Comment:
🗺️ The new `ion_list!`, `ion_sexp!`, and `ion_struct!` macros produce an `Element`. The methods in this file operate on the more specific `List`, `SExp`, and `Struct` types that an `Element` can contain. As a result, we're using the builder APIs rather than the new macros.

Reply:
Is this because the macros are not (or should not) be generic enough to be used recursively to return additional nested lists, structs, or sexp? Should we consider the macros returning an Option of a struct or element?

Reply:
Similar thought should e use `ion_struct!`, `ion_list!` return `Struct`, `List` instead of `Element` and use `ion!` for returning an `Element`? This seems more intuitive to me and may also align with how builder uses `list_builder()`, `sexp_builder()` for building `List` or `SExp`?

Reply:
> Is this because the macros are not (or should not) be generic enough to be used recursively to return additional nested lists, structs, or sexp? [....]

No, `build()` will always successfully return an `Element`. Take a look at the `Value` enum and `Element` struct:

https://github.com/amazon-ion/ion-rust/blob/5c88f8eb712aeee185bc03cbbd4be1171bed4b0a/src/value/owned.rs#L401-L424

`List`, `SExp`, and `Struct` are each variants of the `Value` enum, and an `Element` is a `(Value, annotations)` pair.

In most cases, users will want an `Element` since that's what most APIs expect/return. That's why the macros and the `build()` methods both return an `Element`. However, in some cases (like the code this comment chain is on), you actually want the more specific `List`, `SExp`, or `Struct` types, not the more general `Element` type. In those cases, you can configure your (e.g.) `ListBuilder` and then call `fn build_list(self) -> List` instead of `fn build(self) -> Element`. Does that make sense?

Reply:
> Similar thought should we have ion_struct!, ion_list! return Struct, List instead of Element and use ion! for returning an Element? This seems more intuitive to me and may also align with how builder uses list_builder(), sexp_builder() for building List or SExp?

My aim is to make the most common use cases the most streamlined. Users almost never need/want a `List`/`SExp`/`Struct` after constructing one, and having those types be the default return values for the macros would mean you'd need to do something like:
```rust
let e = Element::from(ion_list![1, 2, 3]);
```
or
```rust
let e: Element = ion_list![1, 2, 3].into();
```
instead of just
```rust
let e = ion_list![1, 2, 3];
```

The macros are named `ion_list!`, `ion_sexp!` and `ion_struct!` because each offers syntax that mirrors the Ion text equivalent.

| Ion | Rust |
|----|------|
|`[1, 2, 3]` | `ion_list! [1, 2, 3]` |
|`(1 2 3)` | `ion_sexp! (1 2 3)` |
|`{foo: "bar"}` | `ion_struct! {"foo": "bar"}`|

Notice that the container macros are intended to make it easy to turn Rust variables into Ion data:
```rust
let first_name = "Lisa";
let last_name = "Jones";
let e = ion_struct! {"first": first_name, "last": last_name};
```
while the `ion!` macro expects an already-valid Ion data stream.  Inserting existing values requires injecting them into Ion data like this:
```rust
let first_name = "Lisa";
let last_name = "Jones";
let ion_data = format!(r#"{{"first": "{}", "last": "{}"}}"#, first_name, last_name);
let e = ion!(ion_data);
```
which is error prone and expensive.

Reply:
> My aim is to make the most common use cases the most streamlined. Users almost never need/want a List/SExp/Struct after constructing one, and having those types be the default return values for the macros would mean you'd need to do something like:

After some more consideration, I ended up swinging the other way on this.

Instead of returning a fully constructed `Element`, the macros now return their corresponding bare value type:
* `ion_list!` returns `List`
* `ion_sexp!` returns `SExp`
* `ion_struct!` returns `Struct`

I did this for a few reasons:
* When working with any other value type (`Decimal`, `Integer`, `bool`, `f64`, etc) it's understood that there's a required conversion step that needs to happen before you can use it as an `Element`. This change makes the bare collection value types consistent with the scalar types; when you build a `List`, `SExp`, or `Struct`, there's now a required conversion step that needs to happen before you can use it as an `Element`.
* Because the bare value types implement `Into<Element>`, you only end up paying the syntactic cost of `.into()` at the top level. For example:
```rust
let list: Element = ion_list! [
  1,
  true,
  ion_struct! {"foo": "bar"}, // nested; like all other bare values in this list, no .into() required
  7.5f64
].into(); // Top level, .into() required
```
* Functions that take an `Into<Element>` instead of an `Element` can accept the bare value types without explicit conversion.
```rust
fn do_something<E: Into<Element>>(value: E) {
  // ...something...
}

do_something(ion_list! [1, 2, 3]); // No explicit conversion required
```
* Thanks to Matt's idea to offer `with_annotations(...)` on values that haven't become `Element`s yet, we can also construct a full `Element` that way:
```rust
let list_element = ion_list! [1, 2, 3].with_annotations(["foo", "bar"]);
```
* There are some methods in the builder APIs where it's nice to be able to easily construct and pass in the bare value types `List`, `SExp`, and `Struct`. While the builder APIs are simple/limited at the moment, I have a sense that we may use that more in the future.